### ISSUE TITLE
Upgrade from 2.3.4 to 5.0.0 fails

### ISSUE THREAD

**Issue by sricci82** [elastic/elasticsearch#21572]
**Elasticsearch version**: 2.3.4

**Plugins installed**: []

**JVM version**: 1.8.0_45-b14

**OS version**: Debian GNU/Linux 8.5 (jessie)

**Description of the problem including expected versus actual behavior**:
Upgrade fails at database start, there seems to be a problem with the mapping of an existing index. Mapping is the following:

```
"mappings": {
    "regtable": {
      "dynamic_templates": [
        {
          "strings": {
            "mapping": {
              "index": "not_analyzed",
              "type": "string"
            },
            "match_mapping_type": "string"
          }
        }
      ],
      "_all": {
        "enabled": false
      }
    }
  },
```

Here's the startup log:

```
[2016-11-15T17:51:00,990][ERROR][o.e.g.GatewayMetaState   ] [elastic-01] failed to read local state, exiting...
java.lang.IllegalStateException: unable to upgrade the mappings for the index [[wireless-2016.06.01/CBIJn2uRQfa3um_SWi9z0g]]
	at org.elasticsearch.cluster.metadata.MetaDataIndexUpgradeService.checkMappingsCompatibility(MetaDataIndexUpgradeService.java:128) ~[elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.cluster.metadata.MetaDataIndexUpgradeService.upgradeIndexMetaData(MetaDataIndexUpgradeService.java:79) ~[elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.gateway.GatewayMetaState.upgradeMetaData(GatewayMetaState.java:248) ~[elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.gateway.GatewayMetaState.<init>(GatewayMetaState.java:91) [elasticsearch-5.0.0.jar:5.0.0]
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[?:?]
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62) [?:1.8.0_45]
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) [?:1.8.0_45]
	at java.lang.reflect.Constructor.newInstance(Constructor.java:422) [?:1.8.0_45]
	at org.elasticsearch.common.inject.DefaultConstructionProxyFactory$1.newInstance(DefaultConstructionProxyFactory.java:49) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.common.inject.ConstructorInjector.construct(ConstructorInjector.java:86) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.common.inject.ConstructorBindingImpl$Factory.get(ConstructorBindingImpl.java:116) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.common.inject.ProviderToInternalFactoryAdapter$1.call(ProviderToInternalFactoryAdapter.java:47) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.common.inject.InjectorImpl.callInContext(InjectorImpl.java:825) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.common.inject.ProviderToInternalFactoryAdapter.get(ProviderToInternalFactoryAdapter.java:43) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.common.inject.Scopes$1$1.get(Scopes.java:59) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.common.inject.InternalFactoryToProviderAdapter.get(InternalFactoryToProviderAdapter.java:50) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.common.inject.SingleParameterInjector.inject(SingleParameterInjector.java:42) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.common.inject.SingleParameterInjector.getAll(SingleParameterInjector.java:66) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.common.inject.ConstructorInjector.construct(ConstructorInjector.java:85) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.common.inject.ConstructorBindingImpl$Factory.get(ConstructorBindingImpl.java:116) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.common.inject.ProviderToInternalFactoryAdapter$1.call(ProviderToInternalFactoryAdapter.java:47) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.common.inject.InjectorImpl.callInContext(InjectorImpl.java:825) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.common.inject.ProviderToInternalFactoryAdapter.get(ProviderToInternalFactoryAdapter.java:43) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.common.inject.Scopes$1$1.get(Scopes.java:59) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.common.inject.InternalFactoryToProviderAdapter.get(InternalFactoryToProviderAdapter.java:50) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.common.inject.InjectorBuilder$1.call(InjectorBuilder.java:191) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.common.inject.InjectorBuilder$1.call(InjectorBuilder.java:183) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.common.inject.InjectorImpl.callInContext(InjectorImpl.java:818) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.common.inject.InjectorBuilder.loadEagerSingletons(InjectorBuilder.java:183) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.common.inject.InjectorBuilder.loadEagerSingletons(InjectorBuilder.java:173) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.common.inject.InjectorBuilder.injectDynamically(InjectorBuilder.java:161) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.common.inject.InjectorBuilder.build(InjectorBuilder.java:96) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.common.inject.Guice.createInjector(Guice.java:96) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.common.inject.Guice.createInjector(Guice.java:70) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.common.inject.ModulesBuilder.createInjector(ModulesBuilder.java:43) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.node.Node.<init>(Node.java:403) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.node.Node.<init>(Node.java:220) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.bootstrap.Bootstrap$5.<init>(Bootstrap.java:191) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.bootstrap.Bootstrap.setup(Bootstrap.java:191) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.bootstrap.Bootstrap.init(Bootstrap.java:286) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.bootstrap.Elasticsearch.init(Elasticsearch.java:112) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.bootstrap.Elasticsearch.execute(Elasticsearch.java:103) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.cli.SettingCommand.execute(SettingCommand.java:54) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.cli.Command.mainWithoutErrorHandling(Command.java:96) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.cli.Command.main(Command.java:62) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:80) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:73) [elasticsearch-5.0.0.jar:5.0.0]
Caused by: java.lang.IllegalArgumentException: Mapper for [_all] conflicts with existing mapping in other types:
[mapper [_all] is used by multiple types. Set update_all_types to true to update [search_analyzer] across all types., mapper [_all] is used by multiple types. Set update_all_types to true to update [search_quote_analyzer] across all types.]
	at org.elasticsearch.index.mapper.FieldTypeLookup.checkCompatibility(FieldTypeLookup.java:130) ~[elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.index.mapper.FieldTypeLookup.copyAndAddAll(FieldTypeLookup.java:94) ~[elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.index.mapper.MapperService.merge(MapperService.java:324) ~[elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.index.mapper.MapperService.merge(MapperService.java:277) ~[elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.cluster.metadata.MetaDataIndexUpgradeService.checkMappingsCompatibility(MetaDataIndexUpgradeService.java:123) ~[elasticsearch-5.0.0.jar:5.0.0]
	... 46 more
```

I think the problem is related to the _all field being disabled, but I checked the cluster with the migration plugin first and it found no issues.

Hope that helps.

Simone.

**Comment by jpountz** [comment]
I think the issue is with `_all` indeed. The error suggests you have multiple types, could you share all your mappings of your index?

**Comment by sricci82** [comment]
Here is it:

{
  "hotspot_stats": {
    "properties": {
      "@timestamp": {
        "format": "strict_date_optional_time||epoch_millis",
        "type": "date"
      },
      "hotspot": {
        "type": "string"
      },
      "count": {
        "type": "long"
      },
      "ap": {
        "type": "string"
      }
    }
  },
  "regtable": {
    "dynamic_templates": [
      {
        "strings": {
          "mapping": {
            "index": "not_analyzed",
            "type": "string"
          },
          "match_mapping_type": "string"
        }
      }
    ],
    "_all": {
      "enabled": false
    },
    "properties": {
      "mac-address": {
        "index": "not_analyzed",
        "type": "string"
      },
      "@timestamp": {
        "format": "strict_date_optional_time||epoch_millis",
        "type": "date"
      },
      "distance": {
        "type": "long"
      },
      "tx-signal-ch0": {
        "type": "long"
      },
      "wireless-comment": {
        "index": "not_analyzed",
        "type": "string"
      },
      "signal-to-noise": {
        "type": "long"
      },
      "rx-signal-ch0": {
        "type": "long"
      },
      "tx-bytes": {
        "type": "long"
      },
      "rx-signal-ch1": {
        "type": "long"
      },
      "tx-signal-ch1": {
        "type": "long"
      },
      "rx-bytes": {
        "type": "long"
      }
    }
  }
}

Just took from the production node (working on 2.3.4).

S.

**Comment by jpountz** [comment]
I tried an upgrade with the above mapping from 2.3.4 to 5.0.1 and it worked so I suspect there is another issue. In which Elasticsearch version was your index created? You can call `curl -XGET localhost:9200/_settings?pretty` to see it, it is stored under the `version.created` key with a cryptic format, but I should be able to translate it.

**Comment by jpountz** [comment]
It is also suspicious that the error complains about a `search_analyzer` while your mappings seem to use the defaults. Can you make sure that this error is not be caused by another index that sets an explicit search_analyzer (or maybe just analyzer)? When you call `curl -XGET localhost:9200/_mappings?pretty` on your 2.x cluster, is there any occurrence of `search_analyzer` or `analyzer`?

**Comment by sricci82** [comment]
```
{
  "index": {
    "creation_date": "1464732109539",
    "number_of_shards": "1",
    "number_of_replicas": "0",
    "uuid": "CBIJn2uRQfa3um_SWi9z0g",
    "version": {
      "created": "2030199"
    }
  }
}
```

```
curl  -sXGET localhost:9200/_mappings?pretty | grep -i 'analyzer'
root@elastic-01:~#
```

Thanks for your time.

Best Regards,
Simone.

**Comment by ywelsch** [comment]
We've found the issue and fixed it in #22236 which has been released as part of ES v5.1.2. Please try the upgrade again using ES v5.1.2.
